cip/1.accepted/CIP2018-08-27-type-filtering-for-IBO.adoc

```diff
@@ -0,0 +1,166 @@
+
+=== Definition
+
+==== Property Type Categories
+Following the link:https://neo4j.com/docs/developer-manual/3.4/cypher/syntax/values/#property-types[property type categorization] of Cypher, we define a _property type category_ as being one of the categories listed in the following table.
+
+|===
+| *Category*      | *Types*
+| Number          | Integer, Float
+| String          | String
+| Boolean         | Boolean
+| Spatial         | Point
+| Temporal        | Date, Time, LocalTime, DateTime, LocalDateTime, Duration
+| ListOfNumbers   | List[Number]
+| ListOfStrings   | List[String]
+| ListOfBooleans  | List[Boolean]
+| ListOfSpatials  | List[Point]
+| ListOfTemporals | List[Date], List[Time], List[LocalTime], List[DateTime], List[LocalDateTime], List[Duration]
+|===
+
+For instance, property type category _Number_ includes all numeric type, precisely _Integer_ and _Float_.
+Property type category _ListOfNumbers_ includes all property types that are a link:https://neo4j.com/docs/developer-manual/3.4/cypher/syntax/lists/[list] where all list elements have a property in the property type category _Number_.
+
+==== Type Filter Predicate
+A _type filter predicate_ (tfp) is an expression of the abstract form _tfp_(_exp_, _ptc_): _Option_[_Boolean_], where _exp_ is an arbitrary expression, _ptc_ is a property type category, and _Option_[_Boolean_] is a nullable _Boolean_, i.e. one of the three values {`true`,`false`,`NULL`}.
+
+==== Type Filter Function
+A _type filter function_ is a predicate function that provides the behavior of a type filter predicate.
+
+=== Proposal
+
+NOTE: We use ⟦_x_⟧ to denote the result of evaluating an expression _x_. For instance, ⟦`3+4`⟧ = `7`.
+
+==== Semantics of Type Filter Predicates
+A type filter predicate _tfp_(_exp_, _ptc_) returns, i.e. ⟦_tfp_(_exp_, _ptc_)⟧ equals
+[horizontal]
+`true`::
+  if ⟦_exp_⟧ is not `NULL` and ⟦_exp_⟧ is of property type _t_ such that _t_ ∈ _ptc_
+`false`::
+  if ⟦_exp_⟧ is not `NULL` and ⟦_exp_⟧ is of property type _t_ such that _t_ ∉ _ptc_
+`NULL`::
+  if ⟦_exp_⟧ is `NULL`
+
+==== Syntax of Type Filter Functions
+We propose to introduce a dedicated type filter function for each of the property type categories.
+Each type filter function is named starting with `is` followed by the of the property type category the function filters for.
+Each type filter function takes an expression (`<exp>`) as its single parameter.
+Precisely, we propose the following type filter function
+
+|===
+| *Category*      | *Type Filter Function*
+| Number          | `isNumber(<exp>)`
+| String          | `isString(<exp>)`
+| Boolean         | `isBoolean(<exp>)`
+| Spatial         | `isSpatial(<exp>)`
+| Temporal        | `isTemporal(<exp>)`
+| ListOfNumbers   | `isListOfNumbers(<exp>)`
+| ListOfStrings   | `isListOfStrings(<exp>)`
+| ListOfBooleans  | `isListOfBooleans(<exp>)`
+| ListOfSpatials  | `isListOfSpatials(<exp>)`
+| ListOfTemporals | `isListOfTemporals(<exp>)`
+|===
```

Comment:
Alternative naming for (e.g.) `isString(x) could be `string.is(x)`. This could be complemented by renaming `toString(x)` to `string.cast(x)`. List functions could perhaps use `string.isList(x)` or even `string.list.is(x)`.

This naming would be consistent with some of the new `date.foo` and `time.foo` functions added alongside the temporal types. It would also allow functions relating to the same type to appear side-by-side in a drop-down list.

Reply:
I kinda like this way of namespacing functions. One comment though is that I'd suggest `list.string.is(x)` for the function that checks if `x` is a `LIST OF STRING`. It's about a list so it makes sense to have that word first.

Reply:
Temporal functions don't use the suggested pattern for type casts. They are of the general form `<type>.<qualifier>(<arguments>)` where a cast is `<type>(<expression>)`: e.g. `datetime('2015-07-21T21:40:32.142+0100')`. To follow the same pattern would indicate that the function for casting to a string would be `string(x)`.

Reply:
SQLite has the same "dynamic typing" (instance vs schema typing) as Cypher. They resolve the sort order issue differently. See https://www.sqlite.org/datatype3.html.

I am interpreting the text of this CIP to mean: "if the result of a function for type X is false then ignore the row concerned in order to ensure that the rows returned have a property which is uniformly of type X or NULL".

If that is a correct understanding, then I believe we should find a way of preserving all the rows in the result set, such that the property concerned is are known to be of type X, or is of another type Y (to be sorted appropriately), or are NULL. This approach would mirror the existing fixed sort order of NULL with respect to null/absent-valued properties, would not require a syntactic change, and would prevent cardinality and value changes when comparing unsorted and sorted result sets.

Reply:
Correcting two mistakes in my comment of yesterday. 

1) SQLite and the Cypher [global sort order](https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2016-06-14-Define-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc) share the same type ordering approach. 

2) But this is irrelevant to the purpose of the proposal, as it aims to impose homogeneous typing by subsetting only those properties of a given type: i.e. the proposal filters out only true results, not true plus NULL.